Adds static and dynamic verification for ConstantDynamic Entry

fengxue-IS opened the pull request:
- Add validation to ldc ldc_w and ldc2_w for ConstantDynamic entry
- Check that the Condy return type matches the bytecode
- Refactor ldc with more informative error messages
- Update description comments on pushLdcType(...)
- Update pushLdcType to expect a constant dynamic to provide a field
- Create new NLS entries for constantDynamic errors

Replaces: #1631
Close: #1278

Signed-off-by: Jack Lu <Jack.S.Lu@ibm.com>

fengxue-IS commented:
@ChengJin01 can you take a look please

fengxue-IS commented:
fixed code format, @ChengJin01 please take a look, thanks

ChengJin01 commented on runtime/bcverify/staticverify.c:
```diff
@@ -257,10 +291,22 @@ checkBytecodeStructure (J9CfrClassFile * classfile, UDATA methodIndex, UDATA len
+			if (CFR_CONSTANT_Dynamic == tag) {
+				J9CfrConstantPoolInfo *constantDynamicSignature = &classfile->constantPool[classfile->constantPool[info->slot2].slot2];
+				if (((flags & BCT_MajorClassFileVersionMask) < BCT_Java11MajorVersionShifted)
+				&& (0 != (flags & BCT_MajorClassFileVersionMask))
```
Better use J9_ARE_ANY_BITS_SET() for (0 != (flags & BCT_MajorClassFileVersionMask)

fengxue-IS replied:
fixed

fengxue-IS commented:
per discussion with @DanHeidinga, removed the zero major version check on constant dynamic as cfdumper no longer sets default classfile version to zero.

ChengJin01 commented:
lgtm

fengxue-IS commented:
senior review: @DanHeidinga please take a look

DanHeidinga commented on runtime/bcverify/staticverify.c:
```diff
@@ -228,26 +228,58 @@ checkBytecodeStructure (J9CfrClassFile * classfile, UDATA methodIndex, UDATA len
+					break;
+				case CFR_CONSTANT_Class:
+					if (((flags & BCT_MajorClassFileVersionMask) < BCT_Java5MajorVersionShifted)
+					&& (J9_ARE_ANY_BITS_SET(flags, BCT_MajorClassFileVersionMask))
```
given the way this has been rewritten, it makes sense to just remove the `&& (J9_ARE_ANY_BITS_SET(flags, BCT_MajorClassFileVersionMask))` as I don't believe it's required anymore.

Previously, cfdump used it.

DanHeidinga commented on runtime/nls/cfre/cfrerr.nls:
```diff
@@ -1361,4 +1361,65 @@ J9NLS_CFR_ERR_CP_ENTRY_INVALID_BEFORE_V55=Constant pool entry not valid in class
+# END NON-TRANSLATABLE
+
+#ldc and ldc_w are not translatable
+J9NLS_CFR_ERR_BC_LDC_NOT_CONSTANT_OR_CONSTANT_DYNAMIC=The ldc and ldc_2 bytecodes must reference a constant or a constant dynamic
```
`constant dynamic`  -> `constant_dynamic`

DanHeidinga commented on runtime/nls/cfre/cfrerr.nls:
```diff
@@ -1361,4 +1361,65 @@ J9NLS_CFR_ERR_CP_ENTRY_INVALID_BEFORE_V55=Constant pool entry not valid in class
+#ldc and ldc_w are not translatable
+J9NLS_CFR_ERR_BC_LDC_NOT_CONSTANT_OR_CONSTANT_DYNAMIC=The ldc and ldc_2 bytecodes must reference a constant or a constant dynamic
+# START NON-TRANSLATABLE
+J9NLS_CFR_ERR_BC_LDC_NOT_CONSTANT_OR_CONSTANT_DYNAMIC.explanation=An 'ldc' or 'ldc_w' bytecode can only reference constant values or a constant dynamic entry.
```
same here

DanHeidinga commented on runtime/nls/cfre/cfrerr.nls:
```diff
@@ -1361,4 +1361,65 @@ J9NLS_CFR_ERR_CP_ENTRY_INVALID_BEFORE_V55=Constant pool entry not valid in class
+# END NON-TRANSLATABLE
+
+#ldc2_w is not translatable
+J9NLS_CFR_ERR_BC_LDC2W_NOT_CONSTANT_OR_CONSTANT_DYNAMIC=The ldc2_w bytecode must reference a constant or a constant dynamic
```
same

DanHeidinga commented on runtime/nls/cfre/cfrerr.nls:
```diff
@@ -1361,4 +1361,65 @@ J9NLS_CFR_ERR_CP_ENTRY_INVALID_BEFORE_V55=Constant pool entry not valid in class
+# END NON-TRANSLATABLE
+
+#ldc and ldc_w are not translatable
+J9NLS_CFR_ERR_BC_LDC_CONSTANT_DYNAMIC_RETURNS_LONG_OR_DOUBLE=Constant dynamic entries referenced by the ldc and ldc_w bytecodes must not return long or double
```
same.  Might be better to use Constant_Dynamic

DanHeidinga commented on runtime/nls/cfre/cfrerr.nls:
```diff
@@ -1361,4 +1361,65 @@ J9NLS_CFR_ERR_CP_ENTRY_INVALID_BEFORE_V55=Constant pool entry not valid in class
+# END NON-TRANSLATABLE
+
+#ldc2_w is not translatable
+J9NLS_CFR_ERR_BC_LDC2W_CONSTANT_DYNAMIC_NOT_LONG_OR_DOUBLE=Constant dynamic entries referenced by the ldc2_w bytecode must return long or double
```
same

fengxue-IS commented:
removed the legacy check condition, changed nls to use `Constant_Dynamic`, also modified some error message wording to make the text more clear to users

AdamBrousseau commented:
Jenkins line endings check

DanHeidinga commented:
The extended test failures are the jdk11 issues we're still working through.  Not related to this PR
